src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -143,151 +131,77 @@ public AssetProvider CreateAssetProvider(Checksum solutionChecksum, SolutionAsse
+            return await _primaryBranchSolutionCache.SlowGetOrCreateSolutionAsync(
+                solutionChecksum,
+                async cancellationToken =>
                 {
-                    var lazySolutionInstance = refCountedLazySolution.TryAddReference();
-                    if (lazySolutionInstance is not null)
-                        return lazySolutionInstance;
-
-                    // Remove the value since it's clearly no longer usable. The cleanupAsync method would have
-                    // removed this value, but has not completed its execution yet.
-                    _solutionChecksumToLazySolution.Remove(solutionChecksum);
-                }
-
-                // We're the first call that is asking about this checksum.  Create a lazy to compute it with a
-                // refcount of 1 (for 'us').
-                refCountedLazySolution = null;
-                var lazySolution = new LazySolution(
-                    getSolutionAsync: cancellationToken => ComputeSolutionAsync(
-                        assetProvider, solutionChecksum, currentSolution, cancellationToken),
-                    cleanupAsync: async () =>
-                    {
-                        // We use CancellationToken.None here as we have to ensure the lazy solution is removed from the
-                        // checksum map, or else we will have a memory leak.  This should hopefully not ever be an issue as we
-                        // only ever hold this gate for very short periods of time in order to set do basic operations on our
-                        // state.
-                        using var _ = await _gate.DisposableWaitAsync(CancellationToken.None).ConfigureAwait(false);
-
-                        // Only remove a value from the map if it still exists and holds the same expected instance
-                        if (_solutionChecksumToLazySolution.TryGetValue(solutionChecksum, out var remainingRefCountedLazySolution)
-                            && remainingRefCountedLazySolution == refCountedLazySolution)
-                        {
-                            _solutionChecksumToLazySolution.Remove(solutionChecksum);
-                        }
-                    });
-                refCountedLazySolution = new ReferenceCountedDisposable<LazySolution>(lazySolution);
-                _solutionChecksumToLazySolution.Add(solutionChecksum, refCountedLazySolution);
-                return refCountedLazySolution;
-            }
+                    var anyBranchSolution = await anyBranchRefCountedSolution.Target.Task.WithCancellation(cancellationToken).ConfigureAwait(false);
+                    var (primaryBranchSolution, _) = await this.TryUpdateWorkspaceCurrentSolutionAsync(workspaceVersion, anyBranchSolution, cancellationToken).ConfigureAwait(false);
+                    return primaryBranchSolution;
+                },
```

Comment:
however, this lambda is different. for the primaryBranch cache we first get hte anyBranch solution, then try to promote it to be the primary branch solution for this workspace.  We then cache that *new* solution instead for this checksum.